src/box/tuple.c

```diff
@@ -128,12 +196,22 @@ tuple_extract_key(const struct tuple *tuple, const struct key_def *key_def,
 	const uint32_t *field_map = tuple_field_map(tuple);
 
 	/* Calculate key size. */
-	for (uint32_t i = 0; i < part_count; ++i) {
+	for (uint32_t i = 0; i < part_count; i++) {
+		uint32_t start = i;
+		for (; i < key_def->part_count - 1; i++) {
+			if (key_def->parts[i].fieldno + 1 !=
+				key_def->parts[i + 1].fieldno) {
+			break;
+			}
+		}
 		const char *field =
 			tuple_field_raw(format, data, field_map,
-					key_def->parts[i].fieldno);
+					key_def->parts[start].fieldno);
 		const char *end = field;
 		mp_next(&end);
+		while (start++ < i) {
+			mp_next(&end);
```

Comment:
This loop can be integrated into `for (; i < key_def->part_count - 1; i++) {` above.